tests/Tests/QuickCheckUtils.hs

```diff
@@ -240,33 +246,36 @@ instance Arbitrary IO.BufferMode where
 --   sometimes contain line endings.)
 -- * Newline translation mode.
 -- * Buffering.
-write_read :: (NFData a, Eq a, Show a)
-           => ([b] -> a)
-           -> ((Char -> Bool) -> a -> b)
-           -> (IO.Handle -> a -> IO ())
-           -> (IO.Handle -> IO a)
-           -> IO.NewlineMode
-           -> IO.BufferMode
-           -> [a]
-           -> Property
-write_read _ _ _ _ (IO.NewlineMode IO.LF IO.CRLF) _ _ = discard
-write_read unline filt writer reader nl buf ts = ioProperty $
-    (===t) <$> act
+write_read :: forall a b c.
+  (Eq a, Show a, Show c, Arbitrary c)
+  => ([b] -> a)
+  -> ((Char -> Bool) -> b -> b)
+  -> (IO.Handle -> a -> IO ())
+  -> (IO.Handle -> IO a)
+  -> (c -> [b])
+  -> [TestTree]
+write_read unline filt writer reader modData
+  = encodings <&> \enc@TextEncoding {textEncodingName} -> testGroup textEncodingName
+    [ testProperty "NoBuffering" $ propTest enc (\() -> IO.NoBuffering) (const False)
+    , testProperty "LineBuffering" $ propTest enc (\() -> IO.LineBuffering) (const False)
+    , testProperty "BlockBuffering" $ propTest enc IO.BlockBuffering (maybe False (< 4))
```

Comment:
Rejection sampling makes the logic convoluted here. Move the sampling of `BufferMode` outside of  `propTest`.

```suggestion
    , testProperty "BlockBuffering" $ forAll (IO.BlockBuffering . fmap ((+ 4) . fromIntegral) <$> (arbitrary :: Gen Word16)) $ \blockBuffering -> propTest enc blockBuffering
```

(note the "`Gen Word16`" was copied from the `Arbitrary IO.BufferMode` instance in this file.)

Reply:
moved to a generator